src/github.com/getlantern/detour/dialer.go

```diff
@@ -0,0 +1,104 @@
+package detour
+
+import (
+	"errors"
+	"fmt"
+	"net"
+	"sync/atomic"
+	"time"
+)
+
+type dialFunc func(network, addr string) (net.Conn, error)
+
+// Dialer returns a function with same signature of net.Dialer.Dial().
+func Dialer(detourDialer dialFunc) func(network, addr string) (net.Conn, error) {
+	return func(network, addr string) (net.Conn, error) {
+		dc := &Conn{
+			addr:            addr,
+			chConnToIOLoop:  make(chan conn),
+			chClose:         make(chan struct{}),
+			chReadRequest:   make(chan ioRequest),
+			chWriteRequest:  make(chan ioRequest),
+			chDialDetourNow: make(chan struct{}),
+			chGetAddr:       make(chan getAddrRequest),
+		}
+
+		type dialResult struct {
+			c   conn
+			err error
+		}
+		ch := make(chan dialResult)
+		dc.expectedConns = 1
+		if !whitelisted(addr) {
+			dc.expectedConns = 2
+		}
+
+		// Dialing logic
+		numDial := int(dc.expectedConns)
+		chLastError := make(chan error, numDial)
+		go func() {
+			if numDial == 2 {
+				go func() {
+					c, err := dialDirect(network, addr)
+					ch <- dialResult{c, err}
+				}()
+				dt := time.NewTimer(DelayBeforeDetour)
+				defer dt.Stop()
+				select {
+				case <-dt.C:
+				case <-dc.chDialDetourNow:
+				}
+				if dc.anyDataReceived() {
+					ch <- dialResult{nil, errors.New("no need to dial detour")}
+					return
+				}
+			}
+			c, err := dialDetour(network, addr, detourDialer)
+			ch <- dialResult{c, err}
+		}()
+
+		// Merge dialing results. Run until all dialing attempts return
+		// but notify caller as soon as any connection available.
+		go func() {
+			var res dialResult
+			for i := 0; i < numDial; i++ {
+				res = <-ch
+				if res.err != nil {
+					atomic.AddUint32(&dc.expectedConns, ^uint32(0))
+					continue
+				}
+				// prevent goroutine from blocking if
+				// I/O loop already exited, or lots of
+				// CLOSE_WAIT connections will be left.
+				select {
+				case dc.chConnToIOLoop <- res.c:
+					chLastError <- nil
+				default:
```

Comment:
Something like this is definitely a bad idea, see https://golang.org/ref/mem#tmp_7. It's a shame that I didn't read doc carefully until now.

I made a simple demo at http://play.golang.org/p/8VJqfmzZnO. By changing the sleep time, sometimes it shows correct result, sometimes it reports deadlock.